### PR TITLE
Optimize date function performance by optimizing locks in DateLUT

### PR DESCRIPTION
**Pull request by liuneng1994**
### Changelog category (leave one):
- Performance Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Optimize DateLUT lock to improve the concurrency performance of date functions

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

10% performance improvement in my test cases
```sql
create table test_date(date String) engine = Memory();

insert into test_date select '2023-02-24 14:53:31' from numbers(30000000);

select fromUnixTimestampInJodaSyntax(parseDateTimeInJodaSyntax(date, 'yyyy-MM-dd HH:mm:ss', 'Europe/Minsk'), 'yyyy-MM-dd HH:mm:ss', 'UTC') from test_date settings max_threads=10 format Null;
```

before:

```
localhost:9000, queries: 100, QPS: 2.046, RPS: 61387601.355, MiB/s: 1639.226, result RPS: 0.000, result MiB/s: 0.000.

0.000%          1.353 sec.      
10.000%         2.117 sec.      
20.000%         2.196 sec.      
30.000%         2.246 sec.      
40.000%         2.331 sec.      
50.000%         2.407 sec.      
60.000%         2.460 sec.      
70.000%         2.524 sec.      
80.000%         2.643 sec.      
90.000%         2.774 sec.      
95.000%         2.922 sec.      
99.000%         3.067 sec.      
99.900%         3.165 sec.      
99.990%         3.165 sec. 
```
after:
```
localhost:9000, queries: 100, QPS: 2.238, RPS: 67132481.680, MiB/s: 1792.631, result RPS: 0.000, result MiB/s: 0.000.

0.000%          1.499 sec.      
10.000%         1.864 sec.      
20.000%         1.972 sec.      
30.000%         2.052 sec.      
40.000%         2.118 sec.      
50.000%         2.153 sec.      
60.000%         2.210 sec.      
70.000%         2.303 sec.      
80.000%         2.437 sec.      
90.000%         2.603 sec.      
95.000%         2.722 sec.      
99.000%         3.108 sec.      
99.900%         3.239 sec.      
99.990%         3.239 sec. 
```

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

---
### Modify your CI run:
**NOTE:** If your merge the PR with modified CI you **MUST KNOW** what you are doing
**NOTE:** Checked options will be applied if set before CI RunConfig/PrepareRunConfig step

#### Include tests (required builds will be added automatically):
- [ ]  Fast test
- [ ]  Integration Tests
- [ ]  Stateless tests
- [ ]  Stateful tests
- [ ]  Unit tests
- [ ]  Performance tests
- [ ]  All with ASAN
- [ ]  All with TSAN
- [ ]  All with Analyzer
- [ ]  Add your option here

#### Exclude tests:
- [ ]  Fast test
- [ ]  Integration Tests
- [ ]  Stateless tests
- [ ]  Stateful tests
- [ ]  Performance tests
- [ ]  All with ASAN
- [ ]  All with TSAN
- [ ]  All with MSAN
- [ ]  All with UBSAN
- [ ]  All with Coverage
- [ ]  All with Aarch64
- [ ]  Add your option here

#### Extra options:
- [ ]  do not test (only style check)
- [ ]  disable merge-commit (no merge from master before tests)
- [ ]  disable CI cache (job reuse)

#### Only specified batches in multi-batch jobs:
- [ ]  1
- [ ]  2
- [ ]  3
- [ ]  4

### CODE REVIEW

**Comment by robot-ch-test-poll** [comment]
*This is an automated comment for commit 87452a07365eefbeecd61a4144e56194d7100ca8 with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/62471/87452a07365eefbeecd61a4144e56194d7100ca8/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>PR Check</td><td>There's no description for the check yet, please add it to tests/ci/ci_config.py:CHECK_DESCRIPTIONS</td><td>✅ success</td></tr>
<tbody>
</table>
</details>

**Comment by liuneng1994** [comment]
In centos7.9, the lock here will cause serious concurrency problems, making the performance close to the performance of single-threaded serial. very strange.

File: src/Common/DateLUT.cpp [hunk 1]
```diff
@@ -182,11 +182,18 @@ DateLUT::DateLUT()
 }
 
 
-const DateLUTImpl & DateLUT::getImplementation(const std::string & time_zone) const
+/// skip thread sanitizer, because the initialization of each time zone data will only happen once.
+/// There should be no serious data race here.
+const DateLUTImpl & DateLUT::getImplementation(const std::string & time_zone) const __attribute__((no_sanitize("thread")))
 {
-    std::lock_guard lock(mutex);
+    // First check without acquiring the lock
```
[the remark this answers is not in the document]
Comment (rschu1ze) [reply src/Common/DateLUT.cpp:189]: Please also use the `NO_SANITIZE_THREAD` macro (base/base/defines.h)

[the remark this answers is not in the document]
Comment (Algunenano) [reply src/Common/DateLUT.cpp:189]: This is a big smell and should not be introduced as is. The tsan warning is valid and should not be silenced.

The number of places that use NO_SANITIZE_THREAD inside CH source should be 0.

[the remark this answers is not in the document]
Comment (Algunenano) [reply src/Common/DateLUT.cpp:189]: BTW, it's fairly obvious why the warning is happening, as reading and writing to an unordered map concurrently is not safe.

[the remark this answers is not in the document]
Comment (liuneng1994) [reply src/Common/DateLUT.cpp:189]: ok, let me try another way

**Comment by alexey-milovidov** [comment]
It will be much better not to call `instance` in a loop.

**Comment by liuneng1994** [comment]
image
I tried other solutions, but they didn't work. This shouldn't be the bottleneck. But I discovered the real reason for the poor concurrency performance of the date parsing function. In the figure, try catch should not be used within the loop for error handling. This will cause serious lock waiting. It seems that the implementation of catch itself has an internal synchronization mechanism. Currently I'm trying to solve this problem.

**Pull request closed without merging**